cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/spark/SparkCqlEvaluator.java

```diff
@@ -538,28 +574,46 @@ protected ContextDefinitions readContextDefinitions(String path) throws Exceptio
      */
-    protected Tuple2<Object, Map<String, Object>> evaluate(Tuple2<Object, List<Row>> rowsByContext,
-            String contextName, CqlEvaluator evaluator, CqlEvaluationRequests requests, SparkOutputColumnEncoder columnEncoder, LongAccumulator perContextAccum) {
+    protected Tuple2<Object, Map<String, Object>> evaluate(Tuple2<Object,
+                                                           List<Row>> rowsByContext,
+                                                           String contextName,
+                                                           CqlEvaluator evaluator,
+                                                           CqlEvaluationRequests requests,
+                                                           SparkOutputColumnEncoder columnEncoder,
+                                                           LongAccumulator perContextAccum,
+                                                           CollectionAccumulator<EvaluationError> errorAccum) {
         perContextAccum.add(1);
         
         List<CqlEvaluationRequest> requestsForContext = requests.getEvaluationsForContext(contextName);
         
         Map<String, Object> expressionResults = new HashMap<>();
         for (CqlEvaluationRequest request : requestsForContext) {
-            try {
-                CqlEvaluationResult result = evaluator.evaluate(request, args.debug ? CqlDebug.DEBUG : CqlDebug.NONE);
-                for (Map.Entry<String, Object> entry : result.getExpressionResults().entrySet()) {
-                    String outputColumnKey = columnEncoder.getColumnName(request, entry.getKey());
-                    expressionResults.put(outputColumnKey, typeConverter.toSparkType(entry.getValue()));
+            for (CqlExpressionConfiguration expression : request.getExpressions()) {
+                CqlEvaluationRequest singleRequest = new CqlEvaluationRequest(request);
+                singleRequest.setExpressions(Collections.singleton(expression));
+                try {
+                    CqlEvaluationResult result = evaluator.evaluate(singleRequest, args.debug ? CqlDebug.DEBUG : CqlDebug.NONE);
+                    for (Map.Entry<String, Object> entry : result.getExpressionResults().entrySet()) {
+                        String outputColumnKey = columnEncoder.getColumnName(request, entry.getKey());
+                        expressionResults.put(outputColumnKey, typeConverter.toSparkType(entry.getValue()));
+                    }
+                } catch (Throwable th) {
+                    if (errorAccum != null) {
+                        Object contextId = rowsByContext._1();
+                        errorAccum.add(new EvaluationError(contextName, contextId, singleRequest.getExpressionNames().iterator().next(), th.getMessage()));
+                    }
+                    else {
+                        throw new RuntimeException(String.format("CQL evaluation failed for ContextName: %s, Request: %s",
```

Comment:
@csandersdev , do you have any thoughts on if the CqlEvaluationRequest itself is ok to include in the message here? I know it has a `conextValue`, but it seems like that's always set to `NA` right now and we don't  do anything  with it. I wasn't sure if this could be used in the future and could ever maybe be set to a context id we wouldn't want to log.

Reply:
Right now, `contextValue` is always user-populated based on what is in the jobs file and we always set that to NA. I think there is a reasonable argument that we should populate that during evaluation with the actual context value, but that isn't something we do today and it wouldn't impact runtime behavior since we group the data ahead of time. The only reason `contextValue` exists in the jobs file at all is because I was trying to model a request object that could be used also in the REST-FHIR path (you'll notice the object lives in cohort-evaluator not cohort-evaluator-spark). It certain creates extra noise in the jobs file and could maybe be dropped. When we added the output column name, it already started to get more Spark-specific than general, so I'm not quite sure what it should be right now.